### PR TITLE
Alek/parameterized test

### LINKED ISSUE #314: Remove duplicates uses of `@TestTemplate` implementations for a single method
See: https://github.com/junit-team/junit5/issues/1224

```java
    @Test
    @RepeatedTest(3)
    @DisplayName("When an entry does not exist, it should be created and initialized to 0")
```

```diff



-- @Test
    @RepeatedTest(3)
    @DisplayName("When an entry does not exist, it should be created and initialized to 0")
```

### PR DESCRIPTION
**Pull request by AlekSimpson**
## What's changed?
Adds a recipe to update `@Test` to `@ParameterizedTest` under certain conditions
 
## What's your motivation?
Closes #314

## Anything in particular you'd like reviewers to focus on?
The import statement for @ParameterizedTest is not being added when I run the tests. I've included the classpath and used all the necessary import methods.

## Anyone you would like to review specifically?
@timtebeek 

### Checklist
- [ ] I've added unit tests to cover both positive and negative cases
- [ ] I've added the license header to any new files through `./gradlew licenseFormat`
- [ ] I've used the IntelliJ auto-formatter on affected files
- [ ] I've updated the documentation (if applicable)

### CODE REVIEW

File: src/test/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotationTest.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    @Test
+    @DocumentExample
+    void replaceTestWithParameterizedTest() {
+        rewriteRun(
+          //language=java
+          java(
+            """
+              import org.junit.jupiter.api.Test;
+              import org.junit.jupiter.params.provider.ValueSource;
+
```
Comment (knutwannheden) [inline src/test/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotationTest.java:32]:
```suggestion
              import static org.junit.jupiter.api.Assertions.*;
```

File: src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    public JavaIsoVisitor<ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            // checks if annotations matches ValueSource, CsvSource or MethodSource
+            public boolean checkForValueAnnotations(J.Annotation ann) {
+                boolean valSourceCheck = (ann.getSimpleName().equals("ValueSource"));
+                boolean csvSourceCheck = (ann.getSimpleName().equals("CsvSource"));
+                boolean methodSourceCheck = (ann.getSimpleName().equals("MethodSource"));
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:36]: You're missing out on matching quite a few other `@...Source` annotations still; see this section of the docs for an overview: https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests

**Comment by AlekSimpson** [comment]
I updated the code with your suggested changes. Two tests are still failing, one fails because there is one extra space added at the start of the lines with the annotations and the method declaration. I'm not sure why that extra space is being included my only guess is it has something to do with the `.addAnnotation()` coordinate line because that is the coordinate that test would be using. 

The other test fails because when I try and reorder the annotations so that the @ParameterizedTest is on top it correctly does that but when it gets converted back into the code the annotations are on the same line (ex: `@ParameterizedTest@SourceValue(..)`).

File: src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                        .javaParser(JavaParser.fromJavaVersion()
+                                .classpathFromResources(ctx, "junit-jupiter-api-5.9", "junit-jupiter-params-5.9"))
+                        .imports("org.junit.jupiter.params.ParameterizedTest")
+                        .build()
+                        .apply(getCursor(), coordinates);
+                maybeAddImport("org.junit.jupiter.params.ParameterizedTest");
+                maybeRemoveImport("org.junit.jupiter.api.Test");
+                List<J.Annotation> ordered = reorderAnnotations(m.getLeadingAnnotations());
+
+                return m.withLeadingAnnotations(ordered);
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:99]: This ought to help your formatting issues, as long as your consistent in your before and after templates in the tests.
```suggestion
                return maybeAutoFormat(m , m.withLeadingAnnotations(ordered), ctx);
```

File: src/test/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotationTest.java [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+                }
+             }
+             """,
+            """
+              import org.junit.jupiter.params.ParameterizedTest;
+              import org.junit.jupiter.params.provider.ValueSource;
+              import static org.junit.jupiter.api.Assertions.*;
+              
+              class TestClass {
+                 @ParameterizedTest
```
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotationTest.java:207]: You're inconsistent in your use of indentation in your tests; here you're indenting with three spaces within the class, and four spaces within the method. You'll likely want to standardize that to not trip up the autoformat.
```suggestion
                  @ParameterizedTest
```

File: src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java (hunk 3 above)
Comment (AlekSimpson) [reply src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:99]: Using `maybeAutoFormat()` helped but the `replaceTestWithParameterizedTestRegardlessOfOrder()`is still failing for formatting issues. Its now adding a newline above the method and two new spaces at the start of each line that the method is on. Seems like the only thing different about this test from the others is that the `reorderAnnotations()` method runs for this test. Could the issue be there maybe?

Comment (timtebeek) [reply src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:99]: You're also somewhat inconsistent with whitespace in your samples; [see below](https://github.com/openrewrite/rewrite-testing-frameworks/pull/353#discussion_r1229268796). I guess clearing that up should help your tests to pass or get closer.

Comment (AlekSimpson) [reply src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:99]: I saw that comment as well and I made sure to go through all the samples and makes sure they all had the same indentation and spacing. Its just this one that is still outputting this weird format. I've also tried using cmd+shift+L on all the samples to see if maybe that would format something I was missing.

**Comment by AlekSimpson** [comment]
With the newest changes I made today all the tests are passing. I'm not sure why the checks on the ci build are still failing.

**Comment by timtebeek** [comment]
> With the newest changes I made today all the tests are passing. I'm not sure why the checks on the ci build are still failing.

You were missing the license headers; I've gone ahead and fixed that in https://github.com/openrewrite/rewrite-testing-frameworks/pull/353/commits/c5c11bbda833a4421033163ba2551fcf8a3652b3 ; it's also part of [the checklist above](https://github.com/openrewrite/rewrite-testing-frameworks/pull/353#issue-1753733083).

File: src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    private static final AnnotationMatcher TEST_ANNOTATION_MATCHER = new AnnotationMatcher("@org.junit.jupiter.api.Test");
+    private static final AnnotationMatcher VALUE_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"ValueSource");
+    private static final AnnotationMatcher CSV_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"CsvSource");
+    private static final AnnotationMatcher METHOD_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"MethodSource");
+    private static final AnnotationMatcher NULL_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"NullSource");
+    private static final AnnotationMatcher EMPTY_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"EmptySource");
+    private static final AnnotationMatcher NULL_EMPTY_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"NullAndEmptySource");
+    private static final AnnotationMatcher ENUM_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"EnumSource");
+    private static final AnnotationMatcher CSV_FILE_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"CsvFileSource");
+    private static final AnnotationMatcher ARG_SOURCE_MATCHER = new AnnotationMatcher(PARAMS_PATH+"ArgumentsSource");
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:42]: Had you also considered something like this?
```suggestion
    private static final List<AnnotationMatcher> SOURCE_ANNOTATIONS = Stream.of(
                    "ArgumentsSource",
                    "CsvFileSource",
                    "CsvSource",
                    "EmptySource",
                    "EnumSource",
                    "MethodSource",
                    "NullAndEmptySource",
                    "NullSource",
                    "ValueSource")
            .map(annotation -> new AnnotationMatcher("@org.junit.jupiter.params.provider." + annotation))
            .collect(Collectors.toList());
```

File: src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    public JavaIsoVisitor<ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+            // checks if annotations matches ValueSource and siblings
+            public boolean checkForSourceAnnotation(J.Annotation ann) {
+                // must be a better way to writer this
+                return (VALUE_SOURCE_MATCHER.matches(ann) || CSV_SOURCE_MATCHER.matches(ann) ||
+                        METHOD_SOURCE_MATCHER.matches(ann) || NULL_SOURCE_MATCHER.matches(ann) ||
+                        EMPTY_SOURCE_MATCHER.matches(ann) || NULL_EMPTY_SOURCE_MATCHER.matches(ann) ||
+                        ENUM_SOURCE_MATCHER.matches(ann) || CSV_FILE_SOURCE_MATCHER.matches(ann) ||
+                        ARG_SOURCE_MATCHER.matches(ann));
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:67]:
```suggestion
                return SOURCE_ANNOTATIONS.stream().anyMatch(matcher -> matcher.matches(ann));
```

File: src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                return ordered;
+            }
+
+            @Override
+            public J.MethodDeclaration visitMethodDeclaration(J.MethodDeclaration md, ExecutionContext ctx) {
+                J.MethodDeclaration m = super.visitMethodDeclaration(md, ctx);
+
+                // return early if @ValueSource and siblings are not detected
+                if (m.getLeadingAnnotations().stream().noneMatch(this::checkForSourceAnnotation) ||
+                    m.getLeadingAnnotations().stream().anyMatch(PARAM_TEST_MATCHER::matches)) {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:89]: I'd flip these two around, as the first one is more likely and easier to test, so probably faster.
```suggestion
                if (m.getLeadingAnnotations().stream().anyMatch(PARAM_TEST_MATCHER::matches) ||
                    m.getLeadingAnnotations().stream().noneMatch(this::checkForSourceAnnotation)) {
```

File: src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+    @Override
+    public String getDescription() {
+        return "Add missing @ParameterizedTest annotation when @ValueSource is used or " +
+               "replace @Test with @ParameterizedTest.";
+    }
+
+    @Override
+    public JavaIsoVisitor<ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:59]: We'll likely want to add a Precondition check here that we're using anything from `org.junit.jupiter.params.provider`, before we evaluate every method anywhere(!)

File: src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+            public List<J.Annotation> reorderAnnotations(List<J.Annotation> annotations) {
+                // if @ParameterizedTest is already at start then return early
+                if (PARAM_TEST_MATCHER.matches(annotations.get(0))) { return annotations; }
+                List<J.Annotation> ordered = new ArrayList<>(annotations);
+                // get @ParameterizedTest
+                List<J.Annotation> filteredList = ordered.stream().filter(PARAM_TEST_MATCHER::matches).collect(Collectors.toList());
+                J.Annotation paramAnn = filteredList.get(0);
+
+                ordered.remove(paramAnn);
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/testing/junit5/AddParameterizedTestAnnotation.java:78]:
```suggestion
                ordered.removeIf(PARAM_TEST_MATCHER::matches);
```

**Review by timtebeek — COMMENTED** [review]
Some quick suggestions for improvement

**Comment by timtebeek** [comment]
Thanks a lot for this effort! I've added a polish commit to simplify things a bit:

- remove `@Test` through `RemoveAnnotationVisitor`, to better reuse edge case handling already present there
- always add `@Parameterized` at the start, removing the need for coordinates.replace() and `reorderAnnotations`
- removes the call to `maybeAutoFormat` which is no longer needed now
- reapplied a formatter to the source code
